symengine/sets.cpp

```diff
@@ -0,0 +1,217 @@
+#include <symengine/sets.h>
+#include <symengine/integer.h>
+
+namespace SymEngine {
+
+Interval::Interval(const RCP<const Number> &start, const RCP<const Number> &end, const bool left_open, const bool right_open)
+    : start_(start), end_(end), left_open_(left_open), right_open_(right_open) {
+        SYMENGINE_ASSERT(is_canonical(start_, end_, left_open_, right_open_));
+    }
+
+bool Interval::is_canonical(const RCP<const Number> &s, const RCP<const Number> &e, bool left_open, bool right_open) const {
+    if (is_a<Complex>(*s) or is_a<Complex>(*e))
+        throw std::runtime_error("Complex set not implemented");
+    if (eq(*e, *s)) {
+        if (left_open or right_open)
+            throw std::runtime_error("Not a valid Interval");
+    } else if (eq(*min({s, e}), *e)){
+        throw std::runtime_error("Not a valid Interval");
+    }
+    return true;
+}
+
+std::size_t Interval::__hash__() const {
+    std::size_t seed = INTERVAL;
+    hash_combine<Basic>(seed, *start_);
+    hash_combine<Basic>(seed, *end_);
+    hash_combine<bool>(seed, left_open_);
+    hash_combine<bool>(seed, right_open_);
+    return seed;
+}
+
+bool Interval::__eq__(const Basic &o) const {
+    if (is_a<Interval>(o)) {
+        const Interval &s = static_cast<const Interval&>(o);
+        return ((this->left_open_ == s.left_open_) and (this->right_open_ == s.right_open_) and
+                eq(*this->start_, *s.start_) and eq(*this->end_, *s.end_));
+    }
+    return false;
+}
+
+int Interval::compare(const Basic &s) const {
+    // compares two interval based on their length
+    SYMENGINE_ASSERT(is_a<Interval>(s))
+    const Interval &o = static_cast<const Interval &>(s);
```

Comment:
@isuruf 
Sorry, but I still do not get this.
Can I do the comparison of `this.end_` and `o.start_` and return result accordingly ?